File: financials/services/departments-service.js
import fetch from "node-fetch";
import pool from "../db/pgPool.js";

class DepartmentsService {

    constructor(url) {
        this.url = url
    }

    async pullData() {
        const res = await fetch(this.url,{
            method: 'GET',
            headers: {
                "x-appsecrettoken": "demo",
                "x-agreementgranttoken": "demo"
            }
        })
        //console.log(await res.json())
        return await res.json();
    }

    async uploadData(data) {
        data.collection.forEach(async(account) => {
            const {
                departmentNumber,
                name,
                distributions
            } = account;
            const insertScript = `INSERT INTO departments(
                number,
                name
                ) VALUES (
                    ${departmentNumber},
                    '${name}'
                )`;
        
            return await pool.query(insertScript,[])

        });
        
    }

    async insertData() {
        const savePullData = await this.pullData()
        const saveUploadData = await this.uploadData(savePullData)
        //return saveUploadData
    }

}

export default DepartmentsService;

File: financials/services/departments_service.py
import json
import requests

from sqlalchemy.orm import Session

from models.department import Department

class DepartmentsService:

    def __init__(self, url, engine):
        self.url = url
        self.engine = engine

    def pull_data(self):
        res = requests.get(self.url,
            headers= {
                "x-appsecrettoken": "demo",
                "x-agreementgranttoken": "demo"
            }
        )
        return json.loads(res.text)

    def upload_data(self,data):
        
        with Session(self.engine) as session:
            for department in data['collection']:
                query = session.query(Department).where(
                    Department.department_number == department['departmentNumber']
                )
                existing_department = query.first()
                if existing_department:
                    existing_department.name = department['name']
                    session.add(existing_department)
                else:
                    new_account = Department(
                        department_number = department['departmentNumber'],
                        name = department['name'],
                    )
                    session.add(new_account)
            session.commit()
            
    def insert_data(self):
        data = self.pull_data()
        self.upload_data(data)
